src/main/java/com/melon/portfoliomanager/services/AssetService.java

```diff
@@ -0,0 +1,60 @@
+package com.melon.portfoliomanager.services;
+
+
+import com.melon.portfoliomanager.dtos.TransactionDto;
+import com.melon.portfoliomanager.exceptions.NoSuchUserException;
+import com.melon.portfoliomanager.models.PortfolioItem;
+import com.melon.portfoliomanager.models.Transaction;
+import com.melon.portfoliomanager.models.TransactionType;
+import com.melon.portfoliomanager.models.User;
+import com.melon.portfoliomanager.repositories.PortfolioItemRepository;
+import com.melon.portfoliomanager.repositories.TransactionRepository;
+import com.melon.portfoliomanager.repositories.UserRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+public class AssetService {
+
+    private TransactionRepository transactionRepository;
+    private PortfolioItemRepository portfolioItemRepository;
+    private UserRepository userRepository;
+
+
+    @Autowired
+    public AssetService(TransactionRepository transactionRepository, PortfolioItemRepository portfolioItemRepository, UserRepository userRepository) {
+        this.transactionRepository = transactionRepository;
+        this.portfolioItemRepository = portfolioItemRepository;
+        this.userRepository = userRepository;
+    }
+
+    public void buyStock(TransactionDto transactionDto) {
+
+        List<User> userList = userRepository.findByUsername(transactionDto.getUsername());
+
+        if (userList.isEmpty()) {
+            throw new NoSuchUserException("There is no user with such username. Enter a valid username.");
+        }
+
+        User user = userList.get(0);
+        Transaction transaction = new Transaction(user, TransactionType.BUY, transactionDto.getAssetSymbol(), transactionDto.getQuantity(), transactionDto.getPrice());
+
+        transactionRepository.save(transaction);
+
+        List<PortfolioItem> pfItemList = portfolioItemRepository.findByUserAndCompanyName(user, transactionDto.getAssetSymbol());
+
+        if (pfItemList.isEmpty()) {
```

Comment:
The code in this if-else block is overcomplicated. I imagine we can do something like this:
- Try to take existing record from database table.
- If no such record, create new entity/domain model of type ```PortfolioItem```.
- If such record exists, then just update the necessary fields.
- Save the entity/domain model to the database. (the ```save``` method will do upsert by default, checking the primary key of the record - if there is a record with the same primary key in the database table, it will update it, otherwise, it will create a new record in the database table)

Reply:
Also, you can extract the dealing with portfolio items in a separate method for better readability.

Reply:
fixed that.